lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <mutex>
+#include <optional>
+#include <thread>
+
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/TracingCommon.h"
+#include "revng/Support/Assert.h"
+
+inline constexpr auto TracingEnv = "REVNG_C_API_TRACE_PATH";
+static bool TracingEnabled = llvm::sys::Process::GetEnv(TracingEnv).hasValue();
+
+// The opposite of a std::recursive_mutex, if locked by the same thread it will
+// assert (this is to avoid a deadlock/malformed output when tracing)
+class NonRecursiveMutex {
+private:
+  std::optional<std::thread::id> ThreadId;
+  std::mutex TheLock;
+
+public:
+  NonRecursiveMutex() {}
+
+  void lock() {
+    if (ThreadId.has_value())
+      revng_assert(std::this_thread::get_id() != *ThreadId,
+                   "NonRecursiveMutex entered twice by the same thread!");
+    TheLock.lock();
+    ThreadId = std::this_thread::get_id();
+  }
+
+  void unlock() {
+    revng_assert(std::this_thread::get_id() == *ThreadId);
+    TheLock.unlock();
+    ThreadId.reset();
+  }
+};
+
+static NonRecursiveMutex TraceLock;
+
+// Helper class for tracing, this will be used by the the argument/return
+// handlers defined below to properly print the value of the arguments onto the
+// YAML tracing file
+class PipelineCTracer {
+private:
+  std::string Path;
+  std::error_code OpenEC;
+  llvm::raw_fd_ostream OS;
+  bool OutputtingArguments;
+
+public:
+  PipelineCTracer(std::string Path) :
```

Comment:
I feel like this class could take a `raw_ostream` instead of a path. The argument for "one day this will be compressed" is suspicious to me. We try hard to flush ASAP to be sure that we don't lose data upon crash. But this idea doesn't play very well with "compressing" stuff on the fly, which AFAIK requires significant buffering.

Reply:
> one day this will be compressed

My idea (which I did not communicate) was that the trace would be compressed some time *after* crashing (e.g. when uploading it to `$STORAGE`) and it would have been a "convenience" being able to run/inspect it without doing `<(xz -ck $file)`
Writing the trace to a compressed file was never (in my mental model of things) never in scope

Reply:
OK, one more reason to make this agnostic from the concept of path and use `raw_ostream` instead.

Reply:
Ok, but if you look [here](https://github.com/revng/revng/commit/6f4d2b36b486daf4e0d7354629c8252e71297f2c#diff-8d2bde1a80c34ad7eb84c0c6f34e25d291b910ae1ee42ca92b27d6e513604465R208) the input to tracewriter is the value of the environment variable. I'm ok to make things agnostic but this would introduce the additional hurdle of storing the ostream outside the class
Do we actually expect to write to something that's not a file?

Reply:
In general, libraries should not consider files. Files are a concept best left to `main` of tools.
Not dealing with files makes it easier to test things.

Reply:
Ok, but here we're inside `librevngPipelineC`, whose job would it be to initialize the `raw_ostream`? `rp_initialize`? But then you'd miss it in the trace.
I can add an additional constructor that takes a "naked" `raw_ostream` but that would bloat the private fields a bit

Reply:
> Ok, but here we're inside librevngPipelineC, whose job would it be to initialize the raw_ostream? 

Whoever is currently calling the constructor needs to be the owner of the `raw_ostream`. So, from what I see, a global variable. Probably a small class that is only constructed as a global variable that owns both `PipelineCTracer` and the stream, which `PipelineCTracer` would only take by reference.